apps/sms/views/shared/js/system_message_handler.js

```diff
@@ -329,6 +338,23 @@
       // Conversation view to the user, we remove that notification from the
       // tray that causes the second system message with "clicked" set to false.
       if (!notification.clicked || !notification.data) {
+        // Force closing the window for notification removal case
+        // because navigation didn't init correctly if pending notification
+        // message existed.
+        if (Navigation.hasPendingInit()) {
+          return Notification.get().then((notifications) => {
```

Comment:
So here's the new approach: Check if we have any other notification. It seems like the notifications we can get would be restricted to the message notification, so I only need to check the number and decide the next step. If we don't have other notification, close app directly, otherwise we check the isNavigating flag and resolve if next toPanel already ongoing/start navigation init if we launch the app directly. I add the onceDocumentIsVisible to make sure it does enter the navigation.toPanel, but it seems not necessary because it would be set even without waiting for visibility. BTW we can also replace the Navigation.init with window.close.